### PR TITLE
fix(r2dbc) #306 - query timeout might be null

### PR DESCRIPTION
**Pull request by oshai**
query timeout might be null

### CODE REVIEW

File: r2dbc-mysql/src/main/java/MysqlConnectionFactoryProvider.kt [hunk 1]
```diff
@@ -67,14 +67,14 @@ class MysqlConnectionFactoryProvider : ConnectionFactoryProvider {
...
-            username = connectionFactoryOptions.getValue(USER) as String,
+            host = connectionFactoryOptions.getValue(HOST) as String? ?: throw IllegalArgumentException("HOST is missing"),
+            port = connectionFactoryOptions.getValue(PORT) as Int? ?: throw IllegalArgumentException("PORT is missing"),
+            username = connectionFactoryOptions.getValue(USER) as String? ?: throw IllegalArgumentException("USER is missing"),
             password = connectionFactoryOptions.getValue(PASSWORD)?.toString(),
             database = connectionFactoryOptions.getValue(DATABASE) as String?,
             applicationName = connectionFactoryOptions.getValue(APPLICATION_NAME) as String?,
             connectionTimeout = (connectionFactoryOptions.getValue(CONNECT_TIMEOUT) as Duration?)?.toMillis()?.toInt() ?: 5000,
-            queryTimeout = connectionFactoryOptions.getValue(QUERY_TIMEOUT) as Duration
+            queryTimeout = connectionFactoryOptions.getValue(QUERY_TIMEOUT) as Duration?
```
Comment (JohnNiang) [inline r2dbc-mysql/src/main/java/MysqlConnectionFactoryProvider.kt:77]: IMO, the option `QUERY_TIMEOUT` is not always of type Duration. For example,  

I will get the following error if I configure r2dbc URL like `r2dbc:mysql://localhost:3306/halo?queryTimeout=10s`:

```java
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class java.time.Duration (java.lang.String and java.time.Duration are in module java.base of loader 'bootstrap')
	at com.github.jasync.r2dbc.mysql.MysqlConnectionFactoryProvider.create(MysqlConnectionFactoryProvider.kt:77) ~[jasync-r2dbc-mysql-2.1.1.jar:na]
	at com.github.jasync.r2dbc.mysql.MysqlConnectionFactoryProvider.create(MysqlConnectionFactoryProvider.kt:23) ~[jasync-r2dbc-mysql-2.1.1.jar:na]
	at io.r2dbc.spi.ConnectionFactories.find(ConnectionFactories.java:110) ~[r2dbc-spi-1.0.0.RELEASE.jar:na]
	at io.r2dbc.spi.ConnectionFactories.get(ConnectionFactories.java:140) ~[r2dbc-spi-1.0.0.RELEASE.jar:na]
	at org.springframework.boot.r2dbc.ConnectionFactoryBuilder$OptionsCapableWrapper.buildAndWrap(ConnectionFactoryBuilder.java:191) ~[spring-boot-3.0.0-M5.jar:3.0.0-M5]
	at org.springframework.boot.r2dbc.ConnectionFactoryBuilder$PoolingAwareOptionsCapableWrapper.buildAndWrap(ConnectionFactoryBuilder.java:204) ~[spring-boot-3.0.0-M5.jar:3.0.0-M5]
	at org.springframework.boot.r2dbc.ConnectionFactoryBuilder.build(ConnectionFactoryBuilder.java:177) ~[spring-boot-3.0.0-M5.jar:3.0.0-M5]
	at org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryConfigurations.createConnectionFactory(ConnectionFactoryConfigurations.java:68) ~[spring-boot-autoconfigure-3.0.0-M5.jar:3.0.0-M5]
	at org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryConfigurations$PoolConfiguration$PooledConnectionFactoryConfiguration.connectionFactory(ConnectionFactoryConfigurations.java:92) ~[spring-boot-autoconfigure-3.0.0-M5.jar:3.0.0-M5]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:130) ~[spring-beans-6.0.0-M6.jar:6.0.0-M6]
	... 114 common frames omitted
```

Comment (oshai) [reply r2dbc-mysql/src/main/java/MysqlConnectionFactoryProvider.kt:77]: Do you know what is the allowed syntax? How it's being parsed?

Comment (JohnNiang) [reply r2dbc-mysql/src/main/java/MysqlConnectionFactoryProvider.kt:77]: I'm not sure what is the exactly allowed syntax. 

But we can refer to [PostgreSQL R2DBC Driver#OptionMapper](https://github.com/pgjdbc/r2dbc-postgresql/blob/d1ecccd5bc0c0338d0b088a1ff0cd97da07c3a7f/src/main/java/io/r2dbc/postgresql/OptionMapper.java#L143-L154) and [its usage](https://github.com/pgjdbc/r2dbc-postgresql/blob/d1ecccd5bc0c0338d0b088a1ff0cd97da07c3a7f/src/main/java/io/r2dbc/postgresql/PostgresqlConnectionFactoryProvider.java#L336).

Or refer to Spring Boot [StringToDurationConverter](https://github.com/spring-projects/spring-boot/blob/fcafd2abdb232e33ab9da1914d0bb59d9fe8398c/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/convert/StringToDurationConverter.java#L45-L50), which supports [SIMPLE](https://github.com/spring-projects/spring-boot/blob/fcafd2abdb232e33ab9da1914d0bb59d9fe8398c/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/convert/DurationStyle.java#L40) and [ISO8601](https://github.com/spring-projects/spring-boot/blob/fcafd2abdb232e33ab9da1914d0bb59d9fe8398c/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/convert/DurationStyle.java#L66) duration style.

Additional references:

- https://en.wikipedia.org/wiki/ISO_8601#Durations

**Comment by oshai** [comment]
Contribution to that is welcome!

**Comment by JohnNiang** [comment]
> Contribution to that is welcome!

I'm so sorry I can't complete this meaningful task because I'm not very familiar with Kotlin. Others who are interested are very welcome to take over ❤️

**Pull request merged**